### PR TITLE
New RPC protocol

### PR DESCRIPTION
**Pull request by ujjboy**
### Ⅰ. Describe what this PR did
- Add new protocol codec.
- Change msgId to `int`
- The response uses the same codec and compressor with the request.

### Ⅱ. Does this pull request fix one issue?
#893

### CODE REVIEW

File: common/src/main/java/io/seata/common/io/UnsafeByteArrayInputStream.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
```
Comment (slievrly) [inline common/src/main/java/io/seata/common/io/UnsafeByteArrayInputStream.java]: should use apache/dubbo license header.

File: common/src/main/java/io/seata/common/io/UnsafeByteArrayOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
```
Comment (slievrly) [inline common/src/main/java/io/seata/common/io/UnsafeByteArrayOutputStream.java]: should use apache/dubbo license header.

File: core/src/main/java/io/seata/core/rpc/netty/AbstractRpcRemoting.java [hunk 3]
```diff
@@ -274,11 +287,13 @@ public void operationComplete(ChannelFuture future) {
      */
     protected void sendRequest(Channel channel, Object msg) {
         RpcMessage rpcMessage = new RpcMessage();
-        rpcMessage.setAsync(true);
-        rpcMessage.setHeartbeat(msg instanceof HeartbeatMessage);
-        rpcMessage.setRequest(true);
+        rpcMessage.setMessageType(msg instanceof HeartbeatMessage ?
+                ProtocolConstants.MSGTYPE_RESQUEST
+                : ProtocolConstants.MSGTYPE_HEARTBEAT);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/AbstractRpcRemoting.java]: mistake？

File: common/src/main/java/io/seata/common/util/CodecUtils.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class CodecUtils {
+
+    /**
+     * 一个byte存两个4bit的信息
+     *
+     * @param b 原始byte
+     * @return byte数组{<16,<16}
+     */
+    public static byte[] parseHigh4Low4Bytes(byte b) {
```
Comment (slievrly) [inline common/src/main/java/io/seata/common/util/CodecUtils.java]: The method name is a bit confusing, remove all Chinese comments.

File: core/src/main/java/io/seata/core/rpc/netty/AbstractRpcRemoting.java [hunk 5]
```diff
@@ -293,17 +308,19 @@ protected void sendRequest(Channel channel, Object msg) {
     /**
      * Send response.
      *
-     * @param msgId   the msg id
+     * @param request  the msg id
      * @param channel the channel
      * @param msg     the msg
      */
-    protected void sendResponse(long msgId, Channel channel, Object msg) {
+    protected void sendResponse(RpcMessage request, Channel channel, Object msg) {
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/AbstractRpcRemoting.java]: just use request.getId why wrap as RpcMessage.   if use RpcMessage ,it is best to wrap msg in the body

File: core/src/main/java/io/seata/core/rpc/ServerMessageListener.java [hunk 6]
```diff
@@ -32,44 +32,44 @@
...
      * On trx message.
      *
-     * @param msgId   the msg id
+     * @param request the msg id
      * @param ctx     the ctx
      * @param message the message
      * @param sender  the sender
      */
-    void onTrxMessage(long msgId, ChannelHandlerContext ctx, Object message, ServerMessageSender sender);
+    void onTrxMessage(RpcMessage request, ChannelHandlerContext ctx, Object message, ServerMessageSender sender);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/ServerMessageListener.java]: RpcMessage contains  message.

File: core/src/main/java/io/seata/core/rpc/ServerMessageListener.java [hunk 7]
```diff
@@ -32,44 +32,44 @@
...
      *
-     * @param msgId            the msg id
+     * @param request          the msg id
      * @param ctx              the ctx
      * @param message          the message
      * @param sender           the sender
      * @param checkAuthHandler the check auth handler
      */
-    void onRegRmMessage(long msgId, ChannelHandlerContext ctx, RegisterRMRequest message,
+    void onRegRmMessage(RpcMessage request, ChannelHandlerContext ctx, RegisterRMRequest message,
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/ServerMessageListener.java]: same as above.

File: core/src/main/java/io/seata/core/rpc/ServerMessageListener.java [hunk 8]
```diff
@@ -32,44 +32,44 @@
...
      *
-     * @param msgId            the msg id
+     * @param request          the msg id
      * @param ctx              the ctx
      * @param message          the message
      * @param sender           the sender
      * @param checkAuthHandler the check auth handler
      */
-    void onRegTmMessage(long msgId, ChannelHandlerContext ctx, RegisterTMRequest message,
+    void onRegTmMessage(RpcMessage request, ChannelHandlerContext ctx, RegisterTMRequest message,
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/ServerMessageListener.java]: same as above

File: core/src/main/java/io/seata/core/rpc/ServerMessageSender.java [hunk 9]
```diff
@@ -31,11 +32,11 @@
     /**
      * Send response.
      *
-     * @param msgId   the msg id
+     * @param request the request
      * @param channel the channel
      * @param msg     the msg
      */
-    void sendResponse(long msgId, Channel channel, Object msg);
+    void sendResponse(RpcMessage request, Channel channel, Object msg);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/ServerMessageSender.java]: same as above

File: core/src/main/java/io/seata/core/protocol/ProtocolConstants.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    public static final byte MSGTYPE_RESQUEST_ONEWAY = 2;
+    public static final byte MSGTYPE_HEARTBEAT = 3;
+
+    // public static final byte MSGTYPE_NEGOTIATOR_REQUEST = 4;
+    // public static final byte MSGTYPE_NEGOTIATOR_RESPONSE = 5;
+
+    // TODO add compressor like CodecFactory
+    public static final byte COMPRESS_NONE = 0;
+
+    public static final byte DEFAULT_CODEC = CodecType.valueOf(ConfigurationFactory.getInstance().getConfig(
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/protocol/ProtocolConstants.java]: not default.

File: core/src/main/java/io/seata/core/protocol/ProtocolConstants.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            ConfigurationKeys.SERIALIZE_FOR_RPC, CodecType.SEATA.name()).toUpperCase()).getCode();
+
+    public static final byte DEFAULT_COMPRESSOR = COMPRESS_NONE;
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/protocol/ProtocolConstants.java]: not default

File: core/src/main/java/io/seata/core/rpc/netty/AbstractRpcRemoting.java [hunk 12]
```diff
@@ -206,10 +219,10 @@ private Object sendAsyncRequest(String address, Channel channel, Object msg, lon
             return null;
         }
         final RpcMessage rpcMessage = new RpcMessage();
-        rpcMessage.setId(RpcMessage.getNextMessageId());
-        rpcMessage.setAsync(false);
-        rpcMessage.setHeartbeat(false);
-        rpcMessage.setRequest(true);
+        rpcMessage.setId(getNextMessageId());
+        rpcMessage.setMessageType(ProtocolConstants.MSGTYPE_RESQUEST_ONEWAY);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/AbstractRpcRemoting.java]: one way?

File: core/src/main/java/io/seata/core/rpc/netty/v1/ProtocolV1Encoder.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    public void encode(ChannelHandlerContext ctx, Object msg, ByteBuf out) {
+        if (out == null) {
+            out = ctx.alloc().buffer();
+        }
+        try {
+            if (msg instanceof RpcMessage) {
+                RpcMessage rpcMessage = (RpcMessage) msg;
+
+                int fullLength = 13;
+                int headLength = 0;
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/v1/ProtocolV1Encoder.java]: headLength is only the size of the headmap?

File: core/src/main/java/io/seata/core/rpc/netty/v1/ProtocolV1Encoder.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+ * @since 0.7.0
+ */
+public class ProtocolV1Encoder extends MessageToByteEncoder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProtocolV1Encoder.class);
+
+    @Override
+    public void encode(ChannelHandlerContext ctx, Object msg, ByteBuf out) {
+        if (out == null) {
+            out = ctx.alloc().buffer();
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/v1/ProtocolV1Encoder.java]: Thread-safe, why the buffer is null?

**Comment by leizhiyuan** [comment]
image

It seems ok in my local machine.

**Pull request merged**